### ISSUE TITLE
Documentation for `modify` confusion

### ISSUE THREAD

**Issue by infinity0** [haskell/vector#371]
I'm a bit confused about the documentation for `modify` in both Data.Vector and Data.Vector.Generic:

~~~
-- | Apply a destructive operation to a vector. The operation will be
-- performed in place if it is safe to do so and will modify a copy of the
-- vector otherwise.
~~~

It's unclear to me how we can possibly detect it is "safe" to perform the operation in-place given that, due to the type signature, the caller is free to pass the input again to another function. Indeed the implementation is just [`modify p = new . New.modify p . clone`](https://github.com/haskell/vector/blob/master/Data/Vector/Generic.hs#L1038) which looks like it does an unconditional copy.

**Comment by infinity0** [comment]
Documentation was originally added in a61207954e75072eb7fd1781e2651b3960d229d3 by @rleshchinskiy

**Comment by lehins** [comment]
@infinity0 Without digging much into the exact proof why this works I will speculate destructive operation can be performed in place thanks to fusion, which is the only way to detect if it is safe to mutate immutable vector or not without relying on linear types.

Here is a rewrite rule that will let us modify a vector without copy, which depends on how it was constructed and if it is being used multiple times or not:

https://github.com/haskell/vector/blob/7ee46342b274042fd033e117316159467574d198/Data/Vector/Generic.hs#L2259-L2260

So if a vector was constructed with a function that relies on `new` then it should be mutated in place without copy.

Whether this works or not as promised at all times or at all can only be confirmed with testing, core inspection and investigating rule firings. Not sure if anything along those lines have been done for `modify` function.

With all this in mind I do agree that documentation is a bit confusing and could use some improvement.

**Comment by infinity0** [comment]
If I understand correctly, the fusion rules only express a *subset* of what safe situations are right? Namely, if the input Vector was itself just created by a call to `new`. However this won't always be the case, especially if the user is using this API with no knowledge of the fusion-specific aspects of Vector - there are lots of other, more common, ways of creating a Vector (e.g. from List) without having used the `new` function.

So the documentation should be made clear regarding that - "Due to fusion RULES, the operation will be performed in place in a subset of the cases where it is safe to do so, namely if its input is statically known to the compiler to have been created with `new`."

**Comment by lehins** [comment]
There are plenty of rewrite rules that will change a pure function that you are using into a call to `new`, so you don't have to manually call `new` yourself in order to hit this optimization. For example the function you mentioned `fromList` will in fact be converted to a call to `new`:

```haskell
fromList xs = new (New.unstream (Bundle.fromList xs))
```

Documenting this will not make things more clear, because it is an optimization. Note that if you compile with `-O0` then destructive operation will never be performed in place because non of the rewrite rules will fire.

I don't think you should take it too seriously and think of this as an optimization that compiler might or might not do it for you. So, documenting that would be good in my opinion: "There is an optimization, which in some case will do a destructive modification safely in place without extra copy"

**Comment by infinity0** [comment]
Yes fair enough, I didn't look into the details of how `fromList` is implemented; however I think my wording was fairly precise, "statically known to the compiler to have been created with `new`" - which covers the cases e.g. if the compiler doesn't know statically if it was created with `fromList` etc.

I can understanding your point that documenting this detail about `new` is too specific however, and simply telling the reader "in some cases [etc]" I agree would be sufficient to communicate the situation clearly (and would solve this issue).

**Comment by Shimuuar** [comment]
Fixed by #466
